src/SimpleGraphs/simpledigraph.jl

```diff
@@ -380,10 +380,15 @@ function copy(g::SimpleDiGraph{T}) where {T<:Integer}
 end
 
 function ==(g::SimpleDiGraph, h::SimpleDiGraph)
-    return vertices(g) == vertices(h) &&
-           ne(g) == ne(h) &&
-           fadj(g) == fadj(h) &&
-           badj(g) == badj(h)
+    if vertices(g) == vertices(h) && ne(g) == ne(h)
+        if fadj(g) == fadj(h) && badj(g) == badj(h)
+            return true
+        else
+            return false
+        end
+    else
+        return false
+    end
```

Comment:
Why would JET report this as an error?

Reply:
Weird, these are all concrete vectors. Also if `==` returns `missing`, I think we still would get an error message.

Maybe one needs to parametrize the method for the inference to correctly work? Something like
``julia
function ==(g::SimpleDiGraph{<:Integer}, h::SimpleDiGraph{<:Integer})
```

Reply:
It's even weirder than that, because the JET error message suggests that all types are correctly inferred:
```julia
   ═════ 2 possible errors found ═════
  ┌ ==(g::Graphs.SimpleGraphs.SimpleDiGraph, h::Graphs.SimpleGraphs.SimpleDiGraph) @ Graphs.SimpleGraphs /home/runner/work/Graphs.jl/Graphs.jl/src/SimpleGraphs/simpledigraph.jl:411
  │ non-boolean `Missing` found in boolean context (1/2 union split): goto %18 if not (Graphs.SimpleGraphs.fadj(g::Graphs.SimpleGraphs.SimpleDiGraph)::Array{Vector{T}, 1} where T<:Integer Graphs.SimpleGraphs.:(==) Graphs.SimpleGraphs.fadj(h::Graphs.SimpleGraphs.SimpleDiGraph)::Array{Vector{T}, 1} where T<:Integer)::Union{Missing, Bool}
  └────────────────────
  ┌ ==(g::Graphs.SimpleGraphs.SimpleDiGraph, h::Graphs.SimpleGraphs.SimpleDiGraph) @ Graphs.SimpleGraphs /home/runner/work/Graphs.jl/Graphs.jl/src/SimpleGraphs/simpledigraph.jl:411
  │ non-boolean `Missing` found in boolean context (1/2 union split): goto %18 if not (Graphs.SimpleGraphs.badj(g::Graphs.SimpleGraphs.SimpleDiGraph)::Array{Vector{T}, 1} where T<:Integer Graphs.SimpleGraphs.:(==) Graphs.SimpleGraphs.badj(h::Graphs.SimpleGraphs.SimpleDiGraph)::Array{Vector{T}, 1} where T<:Integer)::Union{Missing, Bool}
  └────────────────────
```
Basically this tells me that JET knows both forward (or backward) adjacency lists have type `Array{Vector{T}, 1} where T<:Integer`, and yet it can't be arsed to figure out that nothing will go `Missing`

Reply:
Might be this?
```julia
Graphs.SimpleGraphs.badj(h::Graphs.SimpleGraphs.SimpleDiGraph)::Array{Vector{T}, 1} where T<:Integer)::Union{Missing, Bool}
```

Reply:
I wish but JET's tests are not finegrained enough

Reply:
This is because `::(Array{Vector{T}, 1} where T<:Integer) == ::(Array{Vector{T}, 1} where T<:Integer)` is not a fully concrete signature, and there is an implicit recursion within the inference. I have some idea to fix this inference issue, so please watch #541. It should be quick.

Reply:
Ok, I'll do a quick fix like
```julia
if all(fadj(g) .== fadj(h)) && all(badj(g) .== badj(h))
```
but non-allocating and we should be good to merge!

Thank you for your help :)